lib/cast/int32.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+const assert = require('assert');
+const BSON = require('bson');
+
+/**
+ * Given a value, cast it to a Int32, or throw an `Error` if the value
+ * cannot be casted. `null` and `undefined` are considered valid.
+ *
+ * @param {Any} value
+ * @return {Number}
+ * @throws {Error} if `value` does not represent an integer, or is outside the bounds of an 32-bit integer.
+ * @api private
+ */
+
+module.exports = function castInt32(val) {
+  if (val == null) {
+    return val;
+  }
+  if (val === '') {
+    return null;
+  }
+
+  let coercedVal;
+  if (val instanceof BSON.Int32 || val instanceof BSON.Double) {
```

Comment:
we check for `_bsontype` because that lines up with how the serializer interprets our BSON types. `instanceof` is a different contract than our code expects so I think it shouldn't be introduced here so we don't create two versions of what is a true "type check"

Reply:
I agree with using `_bsontype` rather than `instanceof`. Mongoose has an `isBsonType()` helper for this purpose.

The reason why we use `_bsontype` over `instanceof` is that it is common for Node.js apps to have multiple versions of bson floating around, and relying on `instanceof` can lead to hard-to-debug behavior when we get an Int32 instance from a different version of bson module

Reply:
made the change!